README.md
![Icon](https://github.com/netevert/hodl/blob/dev/docs/icon_2.png)
==================================================================
[![Maintenance](https://img.shields.io/maintenance/yes/2018.svg)]()
[![PyPI](https://img.shields.io/pypi/v/hodl.svg)](https://pypi.python.org/pypi/hodl)
[![PyPI](https://img.shields.io/pypi/status/hodl.svg)](https://pypi.python.org/pypi/hodl)
[![PyPI](https://img.shields.io/pypi/pyversions/hodl.svg)](https://pypi.python.org/pypi/hodl)
[![Build Status](https://travis-ci.org/netevert/hodl.svg?branch=master)](https://travis-ci.org/netevert/hodl)
[![Coverage Status](https://coveralls.io/repos/github/netevert/hodl/badge.svg?branch=master)](https://coveralls.io/github/netevert/hodl?branch=master)
[![Donations](https://img.shields.io/badge/donate-bitcoin-orange.svg?logo=bitcoin)](https://github.com/netevert/hodl/blob/master/README.md#donations)

HODL is your friendly, no-nonsense tool to instantaneously check
cryptocurrency prices on the command line, helping you HODL one day at a
time :)


Running HODL is easy:

![Demo](https://github.com/netevert/hodl/blob/master/docs/demo.gif)

Installation
============

To install HODL run:

    pip install hodl

Features
========
Currencies supported:
- Bitcoin ([BTC](https://bitcoin.org/en/))
- Bitcoin Cash ([BCH](https://www.bitcoincash.org/))
- Ethereum ([ETH](https://www.ethereum.org/))
- Litecoin ([LTC](https://litecoin.com/))
- Monero ([XMR](https://getmonero.org/))
- Ripple ([XRP](https://ripple.com/))
- Dash ([DASH](https://www.dash.org/))
- Zcash ([ZEC](https://z.cash/))

<details><summary>View price percentage variations between lookups</summary>
<p>

    C:\>hodl -r
    [*] generating price change report
    1 BTC = 11092.47 USD  -0.07%
    1 BCH = 1243.01 USD   -0.12%
    1 ETH = 849.53 USD    -0.06%
    1 LTC = 208.18 USD     0.00%
    1 XMR = 347.17 USD     0.03%
    1 XRP = 0.92 USD       0.00%
    1 DASH = 599.18 USD   -0.07%
    1 ZEC = 389.29 USD    -0.03%
</p>
</details>

<details><summary>Set preferred conversion currency</summary>
<p>

    C:\>hodl -sf USD
    [*] updating standard fiat to USD ...
    [*] success: USD configured as standard fiat
</p>
</details>

<details><summary>View price for a single cryptocurrency</summary>
<p>

    C:\>hodl -c btc
    1 BTC = 13500.01 USD
    
</p>
</details>

<details><summary>Convert the price for a single cryptocurrency against selected
fiat currencies</summary>
<p>

    C:\>hodl -c ltc -f CHF
    1 LTC = 223.93 CHF
</p>
</details>

<details><summary>Convert the price for a single cryptocurrency against another cryptocurrency</summary>
<p>

    C:\>hodl -cc btc ltc
    1 BTC = 61.36111727 LTC
</p>
</details>

<details><summary>View total value of portfolio holdings</summary>
<p>

    C:\>hodl -vp
    [*] BTC portfolio value: 123475.20 USD
    [*] BCH portfolio value: 234.15 USD
    [*] ETH portfolio value: 672.00 USD
    [*] LTC portfolio value: 120.05 USD
    [*] XMR portfolio value: 1122.43 USD
    [*] XRP portfolio value: 0.00 USD
    [*] DASH portfolio value: 0.00 USD
    [*] ZEC portfolio value: 0.00 USD
</p>
</details>

<details><summary>Configure portfolio holdings</summary>
<p>

    C:\hodl -cp btc 15
    [*] BTC portfolio value set at 15 coins
</p>
</details>
<p></p>

**HODL is under active development**, consult the [GitHub issue tracker](https://github.com/errantbot/hodl/issues)
to check what\'s in the development pipeline.

Bugs/Requests/Contributing
==========================

Please use the [GitHub issue tracker](https://github.com/netevert/hodl/issues) to submit bugs or request
features.

To contribute please consult the [contribution guide](https://github.com/netevert/hodl/blob/master/CONTRIBUTING.md).

License
=======

Distributed under the terms of the [MIT](http://www.linfo.org/mitlicense.html) license, HODL is free and open
source software.

Versioning
==========

This project adheres to [Semantic Versioning](https://semver.org/).

Donations
=========

<details><summary>If you like HODL please consider donating</summary>
<p>
    
    Bitcoin:  13i3hFGN1RaQqdeWqmPTMuYEj9FiJWuMWf
    Litecoin: LZqLoRNHvJyuKz99mNAgVUj6M8iyEQuio9
</p>
</details>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2018-03-11
### Added
- DASH support
- ZEC support
- Improved documentation
- New icon
- News feed

### Changed
- Moved price change report under a dedicated argument (-r)
- Updated demo gif

### Fixed
- Improved speed of price lookups
- Bug that broke the formatting of some crypto to crypto conversion reports

### Removed
- Price change reports from single cryptocurrency price and conversion lookups

## [1.1.2] - 2018-03-01
### Changed
- Project ownership
- Updated homepage links to new project owner
- Updated license references to new project owner

### Fixed
- Broken travis links and setup

## [1.1.1] - 2018-02-28
### Added
- Maintenance status badge on README.md

### Changed
- Updated changelog
- Updated contribution guide

### Fixed
- Fixed bug that caused incorrect price change percentage reports to be
displayed when fiat currency is changed

## [1.1.0] - 2018-02-18
### Added
- Changelog file
- XMR support
- XRP support
- Crypto to crypto exchange rates
- Multi-threading to speed up price retrieval
- Interim project icon
- Donations badge

### Changed
- Colorized output: colorized background in price data for more intuitive reading
- Updated setup.py
- Updated requirements.txt
- Updated demo GIF

### Removed
- Verbose price reports

## [1.0.0] - 2018-01-17
### Added
- Video demo to documentation
- User portfolio
- Colorized output
- Added docstrings
- Improved documentation
- Contribution guide
- Version argument
- Travis continuous integration
- Test suite

### Changed
- Converted readme from .rst to .md
- Updated setup.py

### Fixed
- Align the formatting of the currency exchange rate report
- Increased code coverage to 100%
- Restricted fiat selection to ISO 4217 codes
- Applied minor fixes to setup.py description
- Improved argparse user help menu
- Improved project description in setup.py


tests/test.py
import hodl
import pytest
from tests.utils import create_parser
import configparser as cp


def test_binance_convert_crypto():
    """Tests the output of binance_convert_crypto"""
    out = hodl.binance_convert_crypto(frm="ltc", to="BTC")
    assert "1 LTC =" in out
    assert "BTC" in out
    # test HTTPError handling
    out = hodl.binance_convert_crypto(frm="ABC", to="DEF")
    assert "[*] error, check you are using " \
           "correct crypto symbols" in out


def test_coinbase_convert_crypto():
    """Tests the output of coinbase_convert_crypto"""
    out = hodl.coinbase_convert_crypto(frm="ltc", to="btc")
    assert "1 LTC =" in out
    assert "BTC" in out
    out = hodl.coinbase_convert_crypto(frm="xrp", to="xmr")
    assert "[*] unsupported crypto to crypto conversion operation" in out
    # test HTTPError handling
    out = hodl.coinbase_convert_crypto(frm="ABC", to="DEF")
    assert '[*] unsupported crypto to crypto conversion operation' in out


def test_get_price():
    """Tests the output of get_price()"""
    out = hodl.get_price(crypto="btc", fiat="USD")
    assert "1 BTC =" in out
    assert "USD" in out
    out = hodl.get_price(crypto="xrp", fiat="USD")
    assert "1 XRP =" in out
    assert "USD" in out
    # test HTTPError handling
    out = hodl.get_price(crypto="ABC")
    assert "[*] error, check you are using " \
           "correct crypto and fiat symbols" in out


def test_get_crypto_price():
    """Tests the output of get_crypto_price()"""
    out = hodl.get_crypto_price('xrp', 'usd')
    assert type(out) == float


def test_get_majors():
    """Tests the output of get_majors()"""
    out = hodl.get_majors(fiat="USD")
    assert "1 BTC =" and "USD" in out[0]
    assert "1 BCH =" and "USD" in out[1]
    assert "1 ETH =" and "USD" in out[2]
    assert "1 LTC =" and "USD" in out[3]
    assert "1 XMR =" and "USD" in out[4]
    assert "1 XRP =" and "USD" in out[5]


def test_adjust_readings_to_new_fiat(capfd):
    """Tests the output of adjust_readings_to_new_fiat"""
    hodl.adjust_readings_to_new_fiat('EUR')
    out, err = capfd.readouterr()
    assert "[*] success: EUR configured as standard fiat" in out


def test_set_fiat(capfd):
    """Tests the output and behavior of set_fiat()"""
    config = cp.ConfigParser()
    config.read('../hodl/conf/config.ini')
    backup = config.get("currency", "FIAT")
    hodl.set_fiat(fiat="CHF")
    out, err = capfd.readouterr()
    config.read('../hodl/conf/config.ini')
    assert "[*] updating standard fiat to CHF ..." in out
    assert config.get("currency", "FIAT") == "CHF"
    # return to previous settings and test
    hodl.set_fiat(fiat=backup)
    config.read('../hodl/conf/config.ini')
    assert config.get("currency", "FIAT") == backup


def test_record_data(capfd):
    """Tests the behaviour of record_data()"""
    config = cp.ConfigParser()
    config.read('../hodl/conf/config.ini')
    # get previous config value for readings section
    previous_amount = config.get("readings", "btc")
    test_amount = float(previous_amount) + 100
    # set new value for testing purposes
    hodl.record_data("readings", "btc", str(test_amount))
    config.read('../hodl/conf/config.ini')
    assert config.get("readings", "btc") == str(test_amount)
    # return to previous settings and test
    hodl.record_data("readings", "btc", previous_amount)
    config.read('../hodl/conf/config.ini')
    assert config.get("readings", "btc") == previous_amount
    # get previous config value for portfolio section
    previous_amount = config.get("portfolio", "btc")
    test_amount = float(previous_amount) + 100
    # set new value for testing purposes
    hodl.record_data("portfolio", "btc", test_amount)
    out, err = capfd.readouterr()
    assert "[*] BTC portfolio value set at {} coins".format(test_amount) in out
    config.read('../hodl/conf/config.ini')
    assert config.get("portfolio", "btc") == str(test_amount)
    # return to previous settings and test
    hodl.record_data("portfolio", "btc", previous_amount)
    config.read('../hodl/conf/config.ini')
    assert config.get("portfolio", "btc") == previous_amount
    # test exception handling
    hodl.record_data("portfolio", "btc", "test")
    out, err = capfd.readouterr()
    assert "HODL: error: invalid choice: test (please supply a number)" in out
    hodl.record_data("portfolio", "btc", -1)
    out, err = capfd.readouterr()
    assert "HODL: error: invalid choice: -1 (please supply a positive number)" in out


def test_print_portfolio_value(capfd):
    config = cp.ConfigParser()
    config.read('../hodl/conf/config.ini')
    holding = float(config.get("portfolio", "btc")) * float(config.get("readings", "btc"))
    test_statement = "[*] {} portfolio value: ".format("BTC") + "{0:.2f} USD\n".format(holding)
    hodl.print_portfolio_value("btc")
    out, err = capfd.readouterr()
    assert test_statement == out


def test_print_report(capfd):
    """Tests behavious and output of print_report()"""
    config = cp.ConfigParser()
    config.read('../hodl/conf/config.ini')
    original_amount = config.get("readings", "btc")
    # set testing base amount
    config.set("readings", "btc", str(0))
    with open('../hodl/conf/config.ini', 'w') as configfile:
        config.write(configfile)
    # get previous config value
    base_amount = config.get("readings", "btc")
    test_amount = float(base_amount) + 100
    # set new value for testing purposes
    hodl.record_data("readings", "btc", str(test_amount))
    config.read('../hodl/conf/config.ini')
    assert config.get("readings", "btc") == str(test_amount)
    # test no change report
    hodl.print_report("1 BTC = 100 USD", alignment=1)
    out, err = capfd.readouterr()
    print(out)
    assert "\x1b[37m \x1b[44m" in out  # tests output color is blue
    # test increase report
    hodl.print_report("1 BTC = 120 USD", alignment=1)
    out, err = capfd.readouterr()
    assert "\x1b[37m \x1b[42m" in out  # tests output color is green
    # test decrease report
    hodl.print_report("1 BTC = 80 USD", alignment=1)
    out, err = capfd.readouterr()
    assert "\x1b[37m \x1b[41m" in out  # tests output color is red
    # return to previous settings and test
    hodl.record_data("readings", "btc", original_amount)
    config.read('../hodl/conf/config.ini')
    assert config.get("readings", "btc") == original_amount


def test_main():
    """Tests the output and behaviour of test_main()"""
    parser = create_parser()
    args = parser.parse_args(['-c', 'ltc'])
    assert args.crypto == 'ltc'
    args = parser.parse_args(['-f', 'GBP'])
    assert args.fiat == 'GBP'
    args = parser.parse_args(['-sf', 'USD'])
    assert args.set_fiat == 'USD'
    args = parser.parse_args(['-cp', 'ltc', '2'])
    assert args.configure_portfolio[0] == 'ltc'
    assert args.configure_portfolio[1] == "2"
    # test incorrect values
    with pytest.raises(ValueError):
        parser.parse_args(['-f', 'fake_code'])
    with pytest.raises(ValueError):
        parser.parse_args(['-sf', 'fake_code'])
    with pytest.raises(ValueError):
        parser.parse_args(['-c', 'fake_code'])
    with pytest.raises(ValueError):
        parser.parse_args(["-sf", "USD", "-f", "USD"])


tests/utils.py
import argparse
import hodl


class ErrorRaisingArgumentParser(argparse.ArgumentParser):
    """Custom argument parser which raises a ValueError if
    incorrect value are supplied to choice restricted arguments

    from: https://stackoverflow.com/questions/39028204/using-unittest-to-test-argparse-exit-errors"""
    def error(self, message):
        raise ValueError(message)


def create_parser():
    """Creates a copy of hodl's argument parser for testing purposes"""
    parser = ErrorRaisingArgumentParser()
    group = parser.add_mutually_exclusive_group()
    parser.add_argument("-cp", "--configure_portfolio", help="configure portfolio command",
                        nargs=2, metavar=('CRYPTOCURRENCY', 'AMOUNT'))
    parser.add_argument("-vp", "--view_portfolio", help="view portfolio command", const='all', nargs="?")
    parser.add_argument('-c', '--crypto',
                        help='set the crypto-currency you wish to price check',
                        choices=hodl.cryptos)
    group.add_argument('-f', '--fiat',
                       help='set the fiat currency you wish to use for comparison',
                       choices=hodl.iso4217codes)
    group.add_argument('-sf', '--set_fiat',
                       help='set your preferred fiat currency',
                       choices=hodl.iso4217codes)
    parser.add_argument('-v', '--version', action='version',
                        version='%(prog)s {version}'.format(
                            version=hodl.__version__))
    return parser


hodl.py
#!bin/python
from __future__ import print_function, unicode_literals

try:
    from urllib.request import urlopen, Request, HTTPError
    import configparser as cp
except ImportError:
    import ConfigParser as cp
    from urllib2 import urlopen, Request, HTTPError
import argparse
import json

# constants
description = """
Your friendly, no-nonsense script to instantaneously check cryptocurrency prices, 
helping you HODL one day at a time :)
"""

# read config
config = cp.ConfigParser()
config.read('config.ini')


def get_price(crypto="BTC", fiat=config.get("currency", "FIAT")):
    try:
        url = 'https://api.coinbase.com/v2/prices/{}-{}/spot'.format(crypto, fiat)
        req = Request(url)
        r = urlopen(req).read()
        data = json.loads(r.decode('utf-8'))
        return "1 {} = {} {}".format(data['data']['base'],
                                     data['data']['amount'],
                                     data['data']['currency'])
    except HTTPError:
        return "[*] error, check you are using correct crypto and fiat symbols"


def get_majors(fiat=config.get("currency", "FIAT")):
    return [get_price(crypto, fiat) for crypto in
            ['BTC', 'BCH', 'ETH', 'LTC']]


def set_fiat(fiat):
    try:
        config.set('currency', 'FIAT', fiat)
        with open('config.ini', 'w') as configfile:
            config.write(configfile)
        return "[*] {} configured as standard fiat".format(fiat)
    except Exception as e:
        return "[*] error while configuring fiat, report: ", e


def main():
    parser = argparse.ArgumentParser(description=description)
    group = parser.add_mutually_exclusive_group()
    parser.add_argument('-c', '--crypto',
                        help='the crypto you wish to price check',
                        choices=['BTC', 'BCH', 'ETH', 'LTC'])
    group.add_argument('-f', '--fiat',
                        help='the fiat currency you wish to use for comparison')
    group.add_argument('-sf', '--set_fiat',
                        help='set your preferred fiat currency')
    args = parser.parse_args()

    if args.crypto and args.fiat:
        print(get_price(args.crypto, args.fiat))
    elif (args.set_fiat and args.crypto) or (args.crypto and args.set_fiat):
        print(set_fiat(args.set_fiat))
        print(get_price(args.crypto))
    elif args.crypto:
        print(get_price(args.crypto))
    elif args.set_fiat:
        print(set_fiat(args.set_fiat))
    else:
        for report in get_majors():
            print(report)


if __name__ == '__main__':
    main()


hodl/__init__.py
#!bin/python
from __future__ import print_function, unicode_literals

try:
    from urllib.request import urlopen, Request, HTTPError
    import configparser as cp
except ImportError:
    import ConfigParser as cp
    from urllib2 import urlopen, Request, HTTPError
import argparse
from colorama import init, Fore, Back
import feedparser
from functools import partial
import json
from multiprocessing.pool import ThreadPool
import os
import six

# constants
description = "Your friendly, no-nonsense tool to " \
              "instantaneously check cryptocurrency prices"
epilog = "hodl.py: helping you HODL one day at a time :)"
__version__ = "v.1.2.0"
cryptos = ['btc', 'bch', 'eth', 'ltc', 'xmr', 'xrp', 'dash', 'zec']
iso4217codes = ["AED", "AFN", "ALL", "AMD", "ANG", "AOA", "ARS", "AUD", "AWG",
                "AZN",
                "BAM", "BBD", "BDT", "BGN", "BHD", "BIF", "BMD", "BND", "BOB",
                "BRL",
                "BSD", "BTN", "BWP", "BYN", "BZD", "CAD", "CDF", "CHF", "CLP",
                "CNY",
                "COP", "CRC", "CUC", "CUP", "CVE", "CZK", "DJF", "DKK", "DOP",
                "DZD",
                "EGP", "ERN", "ETB", "EUR", "FJD", "FKP", "GBP", "GEL", "GGP",
                "GHS",
                "GIP", "GMD", "GNF", "GTQ", "GYD", "HKD", "HNL", "HRK", "HTG",
                "HUF",
                "IDR", "ILS", "IMP", "INR", "IQD", "IRR", "ISK", "JEP", "JMD",
                "JOD",
                "JPY", "KES", "KGS", "KHR", "KMF", "KPW", "KRW", "KWD", "KYD",
                "KZT",
                "LAK", "LBP", "LKR", "LRD", "LSL", "LYD", "MAD", "MDL", "MGA",
                "MKD",
                "MMK", "MNT", "MOP", "MRO", "MUR", "MVR", "MWK", "MXN", "MYR",
                "MZN",
                "NAD", "NGN", "NIO", "NOK", "NPR", "NZD", "OMR", "PAB", "PEN",
                "PGK",
                "PHP", "PKR", "PLN", "PYG", "QAR", "RON", "RSD", "RUB", "RWF",
                "SAR",
                "SBD", "SCR", "SDG", "SEK", "SGD", "SHP", "SLL", "SOS", "SPL",
                "SRD",
                "STD", "SVC", "SYP", "SZL", "THB", "TJS", "TMT", "TND", "TOP",
                "TRY",
                "TTD", "TVD", "TWD", "TZS", "UAH", "UGX", "USD", "UYU", "UZS",
                "VEF",
                "VND", "VUV", "WST", "XAF", "XCD", "XDR", "XOF", "XPF", "YER",
                "ZAR",
                "ZMW", "ZWD"]
outlets = {1: ["Crypto Currency News", "https://cryptocurrencynews.com/feed/"]}

# load config file
config_filename = os.path.join(os.path.dirname(os.path.realpath(__file__)),
                               'conf', 'config.ini')
config = cp.ConfigParser()
config.read(config_filename)

# initialise colorama and thread pool
init(autoreset=True)
pool = ThreadPool(processes=4)


def binance_convert_crypto(frm="LTC", to="BTC"):
    """Returns the conversion price from one cypto to another
    using the binance API"""
    try:
        url = "https://api.binance.com/api/v3/ticker/price?symbol={}{}".format(
            frm.upper(), to.upper())
        req = Request(url)
        r = urlopen(req).read()
        data = json.loads(r.decode("utf-8"))
        return "1 {} = {} {}".format(frm.upper(), data["price"], to.upper())
    except HTTPError:
        try:
            url = "https://api.binance.com/api/v3/" \
                  "ticker/price?symbol={}{}".format(to.upper(), frm.upper())
            req = Request(url)
            r = urlopen(req).read()
            data = json.loads(r.decode("utf-8"))
            return "1 {} = {} {}".format(frm.upper(),
                                         1.0 / float(data["price"]), to.upper())
        except HTTPError:
            return "[*] error, check you are using correct crypto symbols"


def coinbase_convert_crypto(frm="LTC", to="BTC"):
    """Returns the conversion price between the supplied crypto
    and fiat currencies using the coinbase API"""
    try:
        if frm in ['btc', 'bch', 'eth', 'ltc'] and \
                to in ['btc', 'bch', 'eth', 'ltc']:
            url = "https://api.coinbase.com/v2/prices/{}-{}/spot"
            req = Request(url.format(frm.upper(), "USD"))
            r = urlopen(req).read()
            data = json.loads(r.decode("utf-8"))
            frm_price = float(data["data"]["amount"])
            req = Request(url.format(to.upper(), "USD"))
            r = urlopen(req).read()
            data = json.loads(r.decode("utf-8"))
            to_price = float(data["data"]["amount"])
            return "1 {} = {} {}".format(frm.upper(),
                                         round(frm_price /
                                               to_price, 8), to.upper())
        else:
            conversion = binance_convert_crypto(frm, to)
            if '[*] error' not in conversion:
                return conversion
            else:
                return "[*] unsupported crypto to crypto conversion operation"

    except (HTTPError, IndexError):
        return "[*] error, check that you are using correct crypto symbols"


def get_price(crypto="BTC", fiat=config.get("currency", "FIAT")):
    """Returns the conversion price between the supplied
    crypto and fiat currencies"""
    try:
        if crypto in ['btc', 'bch', 'eth', 'ltc']:
            url = 'https://api.coinbase.com/v2/prices/{}-{}/spot'.format(
                crypto.upper(),
                fiat)
            req = Request(url)
            r = urlopen(req).read()
            data = json.loads(r.decode('utf-8'))
            return "1 {} = {} {}".format(data['data']['base'],
                                         data['data']['amount'],
                                         data['data']['currency'])
        else:
            return "1 {0} = {1:.2f} {2}".format(crypto.upper(),
                                                get_crypto_price(crypto, fiat),
                                                fiat)
    except (HTTPError, IndexError):
        return "[*] error, check you are using correct crypto and fiat symbols"


def get_crypto_price(crypto, fiat):
    """Helper function to convert any cryptocurrency to fiat"""
    converted_btc_value = float(binance_convert_crypto(
        crypto, "BTC").split('=')[1].strip().split()[0])

    # grab latest bitcoin price
    btc_price = float(get_price("btc", fiat).split('=')[1].strip().split()[0])
    # converted_btc_value * latest reading
    return converted_btc_value * btc_price


def get_majors(fiat=config.get("currency", "FIAT"), fast=False):
    """Returns the conversion prices for all supported crypto-currencies"""
    if not fast:
        reports = pool.map(partial(get_price, fiat=fiat, ), cryptos)
        return reports

    reports = pool.map(partial(get_price, fiat=fiat, ), cryptos)
    for report in reports:
        print_report(report, first_run=True)


def adjust_readings_to_new_fiat(fiat):
    """Helper function to bulk update readings in the configuration file when
    a new default fiat is set"""
    for data in get_majors(fiat):
        record_data('readings',
                    data.split('=')[0].strip().split(' ')[1].lower(),
                    data.split('=')[1].strip().split()[0])
    print("[*] success: {} configured as standard fiat".format(fiat))


def set_fiat(fiat):
    """Sets the default fiat currency for the user"""
    try:
        print("[*] updating standard fiat to {} ...".format(fiat))
        adjust_readings_to_new_fiat(fiat)
        config.set('currency', 'FIAT', fiat)
        with open(config_filename, 'w') as configfile:
            config.write(configfile)
    except Exception as e:
        print("[*] error: while configuring fiat: report: ", e)


def record_data(section, base, amount):
    """Helper function to sanitise and save data to config.ini"""
    try:
        if float(amount) >= 0:
            config.set(section, base, str(amount))
            with open(config_filename, 'w') as configfile:
                config.write(configfile)
            if section == "portfolio":
                print("[*] {} portfolio value set at {} coins".format(
                    base.upper(), amount))
        else:
            print(
                "HODL: error: invalid choice: {} "
                "(please supply a positive number)".format(
                    amount))
    except ValueError:
        print("HODL: error: invalid choice: {} (please supply a number)".format(
            amount))


def print_portfolio_value(base=None):
    """Prints the value of the user's portfolio holdings"""
    portfolio_currency = config.get("currency", "FIAT")
    if base:
        holding = float(config.get("portfolio", base)) * float(
            config.get("readings", base))
        print("[*] {} portfolio value: ".format(base.upper()) +
              "{0:.2f} ".format(holding) +
              "{}".format(portfolio_currency))
    else:
        for base in cryptos:
            holding = float(config.get("portfolio", base)) * float(
                config.get("readings", base))
            print("[*] {} portfolio value: ".format(base.upper()) +
                  "{0:.2f} ".format(holding) +
                  "{}".format(portfolio_currency))


def print_report(report, alignment=0, first_run=False):
    """Prints a crypto-currency exchange rate report"""
    i = alignment - len(report) + 2
    try:
        if "[*] error, check you are using correct " \
           "crypto and fiat symbols" in report:
            print(report)
        else:
            base = report.split("=")[0].split(" ")[1]
            current_amount = float(report.split("=")[1].split(" ")[1])
            # recover cached record
            previous_amount = float(config.get("readings", base.upper()))
            if current_amount > previous_amount and not first_run:
                change = Fore.WHITE + " ".rjust(
                    i) + Back.GREEN + " {0:.2f}%".format(
                    100.0 * current_amount / previous_amount - 100)
                print(report + change)
            elif current_amount < previous_amount and not first_run:
                change = Fore.WHITE + " ".rjust(
                    i) + Back.RED + "{0:.2f}%".format(
                    100.0 * current_amount / previous_amount - 100)
                print(report + change)
            elif current_amount == previous_amount and not first_run:
                change = Fore.WHITE + " ".rjust(i) + Back.BLUE + " 0.00%"
                print(report + change)
            elif first_run:
                print(report)
            record_data("readings", base, str(current_amount))
    except ZeroDivisionError:
        # all readings in the config are set to 0, meaning the user is running
        # the script for the first time. We can thus skip calculating the
        # percentage change in the report by setting first_run to True
        base = report.split("=")[0].split(" ")[1]
        current_amount = float(report.split("=")[1].split(" ")[1])
        record_data("readings", base, str(current_amount))
        print_report(report=report, first_run=True)


def parse_news(outlet):
    """Parses rss news feeds"""
    feed = feedparser.parse(outlet)
    for entry in feed.entries:
        article_title = entry.title
        article_link = entry.link
        article_published_at = entry.published  # Unicode string
        article_author = entry.author
        print("\n==============================")
        print("{}".format(Back.GREEN + article_title))
        print("{}, {}".format(Back.RED + article_author, article_published_at))
        print("{}".format(Back.BLUE + article_link))


def print_news():
    """Function to select and view cryptocurrency news feeds"""
    print("[*] feed listing")
    for i in outlets:
        print("{}: {}".format(Back.RED + str(i), Back.RED + outlets[i][0]))
    outlet_selection = six.moves.input("[*] select feed: ")
    try:
        parse_news(outlets[int(outlet_selection)][1])
    except ValueError:
        print("HODL: error: invalid input: please select a number")


def main():
    """Main program entry point; parses and interprets command line arguments"""
    parser = argparse.ArgumentParser(prog="HODL", description=description,
                                     epilog=epilog)
    group = parser.add_mutually_exclusive_group()
    group_2 = parser.add_mutually_exclusive_group()
    parser.add_argument('-r', '--report',
                        help='prints a price change report', type=bool,
                        const=True, nargs="?")
    parser.add_argument('-n', '--news',
                        help='view news feed', type=bool,
                        const=True, nargs="?")
    group_2.add_argument("-cp", "--configure_portfolio",
                         help="configure portfolio command",
                         nargs=2, metavar=('CRYPTOCURRENCY', 'AMOUNT'))
    group_2.add_argument("-vp", "--view_portfolio",
                         help="view portfolio command", const='all', nargs="?")
    parser.add_argument('-c', '--crypto',
                        help='set the crypto-currency you wish to price check',
                        choices=cryptos)
    group.add_argument('-f', '--fiat',
                       help='set fiat conversion currency',
                       choices=iso4217codes,
                       metavar="ISO 4217 CODE")
    group.add_argument('-sf', '--set_fiat',
                       help='set your preferred fiat currency',
                       choices=iso4217codes,
                       metavar="ISO 4217 CODE")
    parser.add_argument('-v', '--version', action='version',
                        version='%(prog)s {version}'.format(
                            version=__version__))
    parser.add_argument('-cc', '--convert_crypto',
                        help="convert one crypto to another",
                        choices=cryptos,
                        nargs=2)
    args = parser.parse_args()

    if args.crypto and args.fiat:
        print(get_price(args.crypto, args.fiat))
    elif (args.set_fiat and args.crypto) or (args.crypto and args.set_fiat):
        set_fiat(args.set_fiat)
        report = get_price(args.crypto)
        print_report(report, len(report))
    elif args.crypto:
        print(get_price(args.crypto))
    elif args.set_fiat:
        set_fiat(args.set_fiat)
    elif args.configure_portfolio:
        if args.configure_portfolio[0] in cryptos:
            record_data("portfolio", args.configure_portfolio[0],
                        args.configure_portfolio[1])
        else:
            print(
                "HODL: error: argument -cp/--configure_portfolio: invalid choice:"
                " '{}' (choose from {})".format(
                    args.configure_portfolio[0],
                    [str(crypto) for crypto in cryptos]))
    elif args.view_portfolio:
        if args.view_portfolio == "all":
            print_portfolio_value()
        else:
            if args.view_portfolio in cryptos:
                print_portfolio_value(args.view_portfolio)
            else:
                print(
                    "HODL: error: argument -vp/--view_portfolio: invalid choice:"
                    " '{}' (choose from {})".format(
                        args.view_portfolio,
                        [str(crypto) for crypto in cryptos]))
    elif args.convert_crypto:
        if args.convert_crypto[0] in cryptos \
                and args.convert_crypto[1] in cryptos:
            print(coinbase_convert_crypto(args.convert_crypto[0],
                                          args.convert_crypto[1]))
        else:
            print("HODL: error: argument -cc/--convert_crypto: invalid choice:"
                  " '{}' (choose from {})".format(
                args.view_portfolio, [str(crypto) for crypto in cryptos]))

    elif args.report:
        print("[*] generating price change report")
        reports = get_majors()
        i = len(max(reports, key=len))
        for report in reports:
            print_report(report, i)
    elif args.news:
        print_news()
    else:
        get_majors(fast=True)


if __name__ == '__main__':
    main()
